File: python/binary-search.py
# Input: nums = [-1,0,3,5,9,12], target = 9
# Output: 4
# Explanation: 9 exists in nums and its index is 4

class Solution(object):
    def search(self, nums, target):
        low, high = 0, len(nums)-1
        while(low <= high):
            mid = (low+high)//2
            if nums[mid] == target:
                return mid
            elif nums[mid] <= target:
                low = mid+1
            else: 
                high = mid-1
        return -1
    
if __name__ == "__main__":

    solution = Solution()

    nums = [-1,0,3,5,9,12]
    target = 9

    result = solution.search(nums, target)
    print(result)

File: cpp/binary-search.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int low = 0 ;
        int high = nums.size()-1 ;

        while(low<=high){
            int mid = low + (high -low )/2;
            if (nums[mid] == target) {
                return mid ;
            }
            else if(nums[mid] > target ){
                high = mid -1  ;
            }
            else if(nums[mid] < target ){
                low = mid +1 ;
            }
            // mid = low + (high -low )/2;
        }
        return -1 ;
    }
};

int main() {
    Solution myObject; // Create an instance of MyClass

    // Call the Public Function Using the Instance
    vector<int> nums({-1,0,3,5,9,12});
    int target = 5;
    
    cout << myObject.search(nums, target) << "\n";

    return 0;
}